### PR TITLE
Hotfix: Reject degenerate public keys

### PR DESCRIPTION
**Pull request by ffflorian**
This temporarily fixes the rejection of degenerate public keys by throwing an error when a public key equals the point of infinity.

### CODE REVIEW

**Comment by ffflorian** [comment]
As soon as [libsodium.js/update](https://github.com/wireapp/libsodium.js/tree/update) is merged, this hotfix can be reverted.

File: src/proteus/util/ArrayUtil.coffee [hunk 1]
```diff
@@ -105,3 +106,11 @@ module.exports = do ->
       out.push bit_array_to_partial_word(8 * (i & 3), tmp)
     out
 
+  assert_is_not_zeroes: (array) ->
+    not_null = false
+    array.forEach (val) ->
+      if val > 0
+        not_null = true
+
+    if not_null isnt true
```
Comment (lipis) [inline src/proteus/util/ArrayUtil.coffee]: don't use negative variable names..

because the above should be
```
if not not_null
```

Comment (lipis) [inline src/proteus/util/ArrayUtil.coffee]: `is_null` makes more sense

Comment (ffflorian) [inline src/proteus/util/ArrayUtil.coffee]: Thanks, good idea.

File: test/keys/PublicKey.coffee [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+describe 'Public Key', ->
+  it 'should reject shared secrets at the point of infinity', ->
+
+    emptyCurve = new Uint8Array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                                 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
```
Comment (lipis) [inline test/keys/PublicKey.coffee]: Maybe something like that:

```coffeescript
emptyCurve = new Uint8Array([1].concat(0 for [1..30]))
```

(Test it [here](http://coffeescript.org/#try:x%20%3D%20%5B1%5D.concat(0%20for%20%5B1..30%5D)%0A%0Aalert(x)))

File: src/proteus/keys/SecretKey.coffee [hunk 3]
```diff
@@ -23,6 +23,7 @@ sodium = require 'libsodium'
 DontCallConstructor = require '../errors/DontCallConstructor'
 ClassUtil = require '../util/ClassUtil'
 TypeUtil = require '../util/TypeUtil'
+ArrayUtil = require '../util/ArrayUtil'
```
Comment (bennyn) [inline src/proteus/keys/SecretKey.coffee]: Please sort require statements alphabetically.

File: test/keys/PublicKey.coffee [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    a = Proteus.keys.KeyPair.new()
```
Comment (bennyn) [inline test/keys/PublicKey.coffee]: `alice_keypair`...

File: test/keys/PublicKey.coffee [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  it 'should reject shared secrets at the point of infinity', ->
+
+    emptyCurve = new Uint8Array([1].concat(0 for [1..30]))
+
+    alice_keypair = Proteus.keys.KeyPair.new()
+    bob_keypair = Proteus.keys.KeyPair.new()
+
+    assert.deepEqual(
+      alice_keypair.secret_key.shared_secret(bob_keypair.public_key),
+      bob_keypair.secret_key.shared_secret(alice_keypair.public_key))
```
Comment (lipis) [inline test/keys/PublicKey.coffee]:
```coffeescript

assert.deepEqual(
  alice_keypair.secret_key.shared_secret bob_keypair.public_key
  bob_keypair.secret_key.shared_secret alice_keypair.public_key
)

# or

alice_secret = alice_keypair.secret_key.shared_secret bob_keypair.public_key
bob_secret = bob_keypair.secret_key.shared_secret alice_keypair.public_key
assert.deepEqual alice_secret, bob_secret
```

Comment (ffflorian) [inline test/keys/PublicKey.coffee]: Okay 👾

**Pull request merged**